<a href="https://colab.research.google.com/github/ami-null/python-course/blob/main/07_functions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<!-- colab-badge -->
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ami-null/python-course/blob/main/07_functions.ipynb)

# Functions, Lambda, and Type Hints

## Learning Objectives

By the end of this notebook, you should be able to:

- Define and call functions using `def`, with required arguments, named arguments, and default argument values
- Write recursive functions and correctly identify the base case
- Write and apply lambda functions with `map`, `filter`, `sorted`, and `reduce`
- Add type hints to variables and functions

## 1. Functions

A function is a reusable block of code that performs a specific task. Functions eliminate repetition and make code easier to read, test, and maintain.

For example, suppose we have to calculate some permutations and combinations. Each calculation of $^nP_r$ requires calculating factorials two times, and $^nC_r$ require calculating factorials three times. Therefore, instead of repeating the logic of calculating factorial many times, we can create a function that calculates the factorial of an integer and then reuse that function as many times as needed.



### Calling Built-in Functions

Python has many built-in functions.

In [ ]:
# For example, print() is the most common function we have used so far
print("hello")

hello


### Calling a Function with Multiple Arguments

Functions can accept more than one argument. `print()` accepts any number of values and an optional `end` argument that controls what is printed at the very end of the output line.

In [ ]:
name = "Alice"
print("Hello", name)

Hello Alice


In [ ]:
ls1 = [1, 2, 3]

# default end is "\n", so each call goes on its own line
for x in ls1:
    print(x ** 2)

1
4
9


In [ ]:
# setting end=" " keeps output on the same line
for x in ls1:
    print(x ** 2, end=" ")

1 4 9 

### Defining a Function

Functions are defined with the `def` keyword, followed by the function name, its parameters in parentheses, and a colon. The function body must be indented. A `return` statement sends a value back to the caller.

The examples below include **type hints**: the `: int` after each parameter (indicating the data type of the parameters) and the `-> int` (indicating the data type of the return value of the function) before the colon. They state which types the function is meant to work with. They are explained briefly after the next few examples.

In [ ]:
def add_two_things(a: int, b: int) -> int:
    return a + b

def add_three_things(a: int, b: int, c: int) -> int:
    return a + b + c

In [ ]:
add_two_things(5, 6)

11

In [ ]:
add_three_things(1, 2, 3)

6

In [ ]:
# the hints say `int`, but Python does not check them
# the function works with any type that supports the `+` operator
add_two_things("hello", " world")

'hello world'

In [ ]:
# lists support `+` too (concatenation), so this also works despite the `int` hints
add_two_things([5, 6], [7, 8])

[5, 6, 7, 8]

Passing incompatible types to this function would throw an error. Because Python is not checking the types of the two operands before running the code.

In [ ]:
# Python cannot add an int and a str, so this raises a TypeError
# UNCOMMENT THE FOLLOWING LINE TO SEE THE ERROR:

# add_two_things(5, "abc")

### A Brief Introduction to Type Hints

The functions above carry **type hints**. `a: int` says that parameter `a` is expected to be an `int`, and `-> int` says that the function is expected to return an `int`. The same pattern works for other types, for example `name: str`, `score: float`, or `-> bool`. Collections and alternatives are written in a similar way: `list[float]` means a list of floats, and `int | str` means an `int` or a `str`. Section 4 covers these in detail.

**Type hints are not enforced.** Python ignores them when the code runs, which is why `add_two_things("hello", " world")` and `add_two_things([5, 6], [7, 8])` worked above even though the hints say `int`. Hints document the intended types for readers, editors, and static analysis tools. To actually reject the wrong types at runtime, the function has to check them itself, which is what the next section does.

### Checking and Enforcing Data Type in a Function
`isinstance(object, type)` checks whether an object is of a given type. It can be used to enforce that a function only operates on the intended types.

In [ ]:
def add_two_integers(a: int, b: int) -> int | str:
    if isinstance(a, int) and isinstance(b, int):
        return a + b
    else:
        return "Please provide two integers"

In [ ]:
print(add_two_integers(3, "four"))
print(add_two_integers(3, 4))

Please provide two integers
7


### Named Arguments

Arguments can be passed by name. This makes calls more readable and allows arguments to be passed in any order.

In [ ]:
def show_args(n: int, r: int) -> str:
    return f"called with n={n}, r={r}"

# positional - order matters
print(show_args(5, 3))

# named - order does not matter
print(show_args(r=3, n=5))

called with n=5, r=3
called with n=5, r=3


### Examples

#### Example: Factorial

In [ ]:
def fact(n: int) -> int:
    if n < 0:
        raise ValueError("n must be non-negative")
    result = 1
    while n > 1:
        result *= n
        n -= 1
    return result

In [ ]:
fact(5)

120

#### Example: Permutation, $^nP_r$

`permutation` reuses `fact`, which is one reason to write reusable functions: you build larger pieces from smaller ones which helps breaking down a complex problem into smaller manageable problems.

In [ ]:
def permutation(n: int, r: int) -> int:
    return fact(n) // fact(n - r)

In [ ]:
print(permutation(5, 3))
print(permutation(n=5, r=3))

60
60


#### Example: GCD

A common mistake when writing this function is naming a local variable the same as the function itself, which shadows the function name and causes unexpected behaviour. The version below uses `divisor` as the local variable name to avoid this.

In [ ]:
def gcd(a: int, b: int) -> int:
    divisor = min(a, b)
    while a % divisor != 0 or b % divisor != 0:
        divisor -= 1
    return divisor

In [ ]:
gcd(25, 15)

5

#### Example: Prime Check

In [ ]:
def is_prime(n: int) -> bool:
    if n < 2:
        return False
    for i in range(2, n):
        if n % i == 0:
            return False
    return True

In [ ]:
print(is_prime(7))
print(is_prime(10))
print(is_prime(1))

True
False
False


#### Gradebook: Computing an Average Score

In [ ]:
def compute_average(scores: list[float]) -> float:
    return sum(scores) / len(scores)

compute_average([85, 90, 78])

84.33333333333333

#### Gradebook: Assigning a Letter Grade

In [ ]:
def letter_grade(score: float) -> str:
    if score >= 90:
        return "A"
    elif score >= 80:
        return "B"
    elif score >= 70:
        return "C"
    elif score >= 60:
        return "D"
    else:
        return "F"

In [ ]:
print(letter_grade(92))
print(letter_grade(83))
print(letter_grade(55))

A
B
F


### Recursive Functions

A recursive function calls itself within its own body. Every recursive function needs a **base case**: a condition under which it stops calling itself. Without a base case the function recurses indefinitely and causes a `RecursionError`.

#### Example: Calculating Factorial using Recursion
The factorial can be defined recursively as:

$$
f(n) =
\begin{cases}
1, & \text{if } n = 0 \\[6pt]
n \times f(n - 1), & \text{if } n > 0
\end{cases}
$$

In [ ]:
def fact_recurs(n: int) -> int:
    if n < 0:
        raise ValueError("n must be non-negative")
    if n == 0:    # base case: stops the recursion
        return 1
    else:
        return n * fact_recurs(n - 1)

fact_recurs(5)

120

### Exercises - Section 1

**Exercise 1**

Write a function `is_even(n)` that returns `True` if `n` is even and `False` otherwise. Include type hints for the parameter and the return value.

In [ ]:
# your code here

**Exercise 2**

Write a function `celsius_to_fahrenheit(c)` that converts a Celsius temperature to Fahrenheit. Include type hints for the parameter and the return value.

$$F = \frac{9}{5}\,C + 32$$

In [ ]:
# your code here

**Exercise 3**

Write a function `grade_report(name, scores)` where `name` is a string and `scores` is a list of exam scores. The function should perform the following:

1. Computes the average using `compute_average()` defined in this notebook
2. Assigns a letter grade using `letter_grade()` defined in this notebook
3. Returns a formatted string such as `"Alice | Average: 84.33 | Grade: B"`

Use an f-string. Round the average to two decimal places with `round(value, 2)`. Include type hints for the parameters and the return value.

In [ ]:
# your code here

## 2. Default Values for Arguments

A parameter can have a default value. When the caller does not supply that argument, the default is used.

**Parameters with defaults must come after parameters without defaults in the function signature.**

In [ ]:
# since `exponent` has a default value, it must come after the
# argument without default value which is `base`
def power(base: float, exponent: float = 2) -> float:
    return base ** exponent

In [ ]:
print(power(5))       # uses default exponent of 2
print(power(2, 10))   # overrides the default

25
1024


#### Gradebook: Pass/Fail with a Configurable Threshold

The default threshold is 50, but it can be overridden for courses with a different passing cutoff.

In [ ]:
def is_passing(score: float, threshold: float = 50) -> bool:
    return score >= threshold

In [ ]:
print(is_passing(45))                # fails with default threshold
print(is_passing(45, threshold=40))  # passes when threshold is lowered
print(is_passing(70))                # passes with default threshold

False
True
True


**Exercise 4**

Write a function `repeat_string(s, times=2)` that returns the string `s` repeated `times` times.
For example, `repeat_string("go", 3)` should return `"gogogo"` and `repeat_string("hi")` should return `"hihi"`. Include type hints for the parameters and the return value.

In [ ]:
# your code here

**Exercise 5 (builds on Exercise 3)**

Modify `grade_report()` from Exercise 3 and create a new function named `grade_report_2()` so that the `name` argument has a default value of `"Student"`. Demonstrate using the updated function twice:

1. With both a name and a list of scores
2. With only a list of scores, relying on the default name

Note that, you will have to make the `name` the second argument and `scores` the first argument, because arguments with default values must come after arguments without default values. Include type hints.

In [ ]:
# your code here

## 3. Lambda Functions

Lambda functions are small, anonymous functions defined without a name. They are written using the `lambda` keyword and are typically used as short-lived arguments to other functions.

In other languages, equivalent constructs appear under different names: arrow functions in JavaScript, anonymous functions in R, and lambda expressions in Java and C++.

- **When to use `lambda`:** short, one-line operations passed directly into another function.

- **When to use `def`:** anything that requires a name, multiple lines, or is reused in more than one place.

### Syntax and Basic Definition

In [ ]:
# named function version
def square_fn(x: int) -> int:
    return x ** 2

# equivalent lambda - assigned to a variable here only for demonstration
# lambdas are usually not given names like this
square = lambda x: x ** 2

print(square_fn(4))
print(square(4))

16
16


### Lambda with Multiple Arguments

In [ ]:
add = lambda a, b: a + b
print(add(3, 7))

format_name = lambda first, last: f"{last}, {first}"
print(format_name("Alan", "Turing"))

10
Turing, Alan


### Common Use Cases

Lambda functions are most useful with Python's built-in higher-order functions (higher-order functions are functions which accept other functions as arguments).

#### a. `map()`: Apply a function to each element

In [ ]:
nums = list(range(1, 11))
nums

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [ ]:
squared = list(map(lambda x: x ** 2, nums))
print(squared)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [ ]:
words = ["python", "java", "perl", "javascript", "php"]
upper_words = list(map(lambda w: w.upper(), words))
print(upper_words)

['PYTHON', 'JAVA', 'PERL', 'JAVASCRIPT', 'PHP']


#### b. `filter()`: Keep elements that satisfy a condition

In [ ]:
evens = list(filter(lambda x: x % 2 == 0, nums))
print(evens)

[2, 4, 6, 8, 10]


#### c. `sorted()`: Custom sort key

`sorted()` accepts a `key` argument: a function applied to each element before comparison. The elements themselves are not modified, only the comparison order.

In [ ]:
people = [
    {"name": "Alice", "age": 30},
    {"name": "Bob",   "age": 25},
    {"name": "Eve",   "age": 35},
]

In [ ]:
# ascending by age
print(sorted(people, key=lambda p: p["age"]))

[{'name': 'Bob', 'age': 25}, {'name': 'Alice', 'age': 30}, {'name': 'Eve', 'age': 35}]


In [ ]:
# descending by age
print(sorted(people, key=lambda p: p["age"], reverse=True))

[{'name': 'Eve', 'age': 35}, {'name': 'Alice', 'age': 30}, {'name': 'Bob', 'age': 25}]


#### d. `reduce()`: Combine elements cumulatively

`reduce(f, [a, b, c, d])` computes `f(f(f(a, b), c), d)`. It reduces a sequence to a single value by repeatedly applying the function to the accumulated result and the next element.

In [ ]:
nums

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [ ]:
from functools import reduce

In [ ]:
total = reduce(lambda x, y: x + y, nums)
print(total)    # equivalent to sum(nums)

55


In [ ]:
product = reduce(lambda x, y: x * y, nums)
print(product)

3628800


#### Gradebook: Ranking Students by Average Score

In [ ]:
students = [
    {"name": "Alice", "scores": [85, 90, 78]},
    {"name": "Bob",   "scores": [72, 68, 74]},
    {"name": "Carol", "scores": [91, 95, 88]},
    {"name": "Dave",  "scores": [55, 60, 58]},
]

In [ ]:
ranked = sorted(
    students,
    key=lambda s: compute_average(s["scores"]),
    reverse=True
)

ranked

[{'name': 'Carol', 'scores': [91, 95, 88]},
 {'name': 'Alice', 'scores': [85, 90, 78]},
 {'name': 'Bob', 'scores': [72, 68, 74]},
 {'name': 'Dave', 'scores': [55, 60, 58]}]

### Exercises - Section 3

**Exercise 6 (builds on Exercise 2)**

You have a list of Celsius temperatures: `[0, 20, 37, 100]`.

Use `map` and the `celsius_to_fahrenheit()` function from Exercise 2 to convert the entire list to Fahrenheit. Print the result as a list.

In [ ]:
temps_c = [0, 20, 37, 100]
# your code here

**Exercise 7**

Given the student list below, complete the following steps in sequence:

1. Use `filter` and a lambda to keep only students whose average score (computed with `compute_average()` from the lecture) is above 50.
2. Use `sorted` and a lambda to rank the passing students by average score, highest first.
3. Print each passing student's name and average score rounded to two decimal places.

In [ ]:
students = [
    {"name": "Alice", "scores": [85, 90, 78]},
    {"name": "Bob",   "scores": [40, 35, 50]},
    {"name": "Carol", "scores": [70, 65, 80]},
    {"name": "Dave",  "scores": [20, 30, 25]},
]

# your code here

## 4. Type Hints

Python is dynamically typed: you do not need to declare variable types. Type hints, introduced in Python 3.5 (PEP 484), let you annotate variables and functions with expected types. You have already used them on function parameters and return values in the earlier sections; this section covers them in full.

**Python ignores type hints at runtime.** They exist for developers, editors, and static analysis tools (like `mypy`).

Reasons to use them:

- Makes code easier to read and understand at a glance
- Enables better autocomplete and error detection in editors
- Allows static type checkers like `mypy` to catch type mismatches before the code runs

### Annotating Variables

In [ ]:
name: str = "Alice"
age: int = 25
height: float = 5.6
is_enrolled: bool = True

### Annotating Functions

As in the earlier examples, the return type is written after `->` in the function signature.

In [ ]:
def multiply(x: int, y: int) -> int:
    return x * y

def greet(name: str) -> str:
    return f"Hello, {name}!"

print(multiply(3, 4))
print(greet("Alice"))

12
Hello, Alice!


### Union Types with `|`

Since Python 3.10, the `|` operator expresses that a value can be one of several types. Use `X | None` when a value may be absent. `add_two_integers` earlier returned either an `int` or a `str`, which is why its return type was written `int | str`.

In [ ]:
def add(a: int | float, b: int | float) -> int | float:
    return a + b

print(add(3, 4.5))

7.5


In [ ]:
def find_student(name: str) -> dict | None:
    # returns a student record if found, None otherwise
    db = {"Alice": {"score": 85}, "Bob": {"score": 72}}
    return db.get(name)

print(find_student("Alice"))
print(find_student("Eve"))

{'score': 85}
None


### Type Annotations for Collections (lists, tuples, dictionaries etc.)

In [ ]:
scores: list[int] = [85, 90, 78]    # it means that score is a list of integers

# student is a dictionary, the keys are strings and the values are either string or integers:
student: dict[str, str | int] = {"name": "Alice", "age": 25}

# coordinates is a tuple of two integers
coordinates: tuple[int, int] = (10, 20)

# tags is a set of strings
tags: set[str] = {"python", "typing"}

# nickname is a single value, either a string or None
# None conveys to other developers that nickname is optional
nickname: str | None = None

### What a Type Checker Would Flag

Type hints are not enforced at runtime, but a static type checker like `mypy` would flag mismatches before the code runs:

```python
def add(a: int, b: int) -> int:
    return a + b

result = add(5, "3")
# mypy error: Argument 2 to "add" has incompatible type "str"; expected "int"
```

Running this code would also raise a `TypeError`, but a type checker can identify this issue before the code is executed.